suite/tests/client-interface/drbbdup-empty-bb-test.dll.c

```diff
@@ -0,0 +1,300 @@
+/* **********************************************************
+ * Copyright (c) 2022 Google, Inc.   All rights reserved.
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL GOOGLE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* Tests the drbbdup extension. */
+
+#include "dr_api.h"
+#include "client_tools.h"
+#include "drmgr.h"
+#include "drbbdup.h"
+
+#define TEST_PRIORITY_APP2APP_NAME "TEST_PRIORITY_APP2APP"
+#define TEST_PRIORITY_APP2APP 100
+
+#define ORIG_ANALYSIS_VAL (void *)555
+#define ANALYSIS_VAL_1 (void *)888
+#define ANALYSIS_VAL_2 (void *)999
+
+static bool orig_analysis_called = false;
+static bool orig_analysis_destroy_called = false;
+static bool default_analysis_called = false;
+static bool case1_analysis_called = false;
+static bool case1_analysis_destroy_called = false;
+static bool case2_analysis_called = false;
+static bool case2_analysis_destroy_called = false;
+static bool instrum_called = false;
+static bool encountered_empty = false;
+static bool is_cur_empty = false;
+
+/* Assume single threaded. */
+static uintptr_t encode_val = 3;
+
+static bool
+is_empty_bb(instrlist_t *bb)
+{
+    return instrlist_first(bb) == NULL;
+}
+
+static dr_emit_flags_t
+remove_app_instr(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
+                 bool translating)
+{
+    instr_t *first_instr = instrlist_first_app(bb);
+    CHECK(instr_get_next(first_instr) == NULL, "must just be 1 instr");
```

Comment:
It took a while to understand how this works: this relies on passing -max_bb_instrs 0.  Please add a comment about that, maybe here and at the top of the file.

Reply:
Sorry, about that! Will add a comment to make it clearer.

Reply:
Done.